File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    visit = [0 for _ in range(len(dungeons))]
    global answer
    answer = 0
    
    def dfs(k, cnt):
        global answer
        answer = max(cnt, answer)
        
        for i in range(len(dungeons)):
            if k >= dungeons[i][0] and visit[i] == 0:
                visit[i] = 1
                dfs(k-dungeons[i][1], cnt+1)
                visit[i] = 0
            
    dfs(k, 0)
    return answer
    
'''
모든 던전을 for로 검색하여, 갈 수 있는 경로를 다 가보기
탐색 이후에는 방문 처리를 다시 되돌린다.
'''

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool visited[8] = {false,};
int maxCount = 0;

void dfs(int k, vector<vector<int>> dungeons, int count) {
    maxCount = max(maxCount, count);
    
    for (int i=0; i<dungeons.size(); i++) {
        if (!visited[i] && k >= dungeons[i][0]){
            visited[i] = true;
            dfs(k-dungeons[i][1], dungeons, count + 1);
            visited[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    dfs(k, dungeons, 0);
    return maxCount;
}